<launch>

  <arg name="ouster_ns" default="ouster" doc="Override the default namespace of all ouster nodes"/>
  <arg name="metadata" default ="" doc="path to read metadata file when replaying sensor data"/>
  <arg name="bag_file" default="" doc="file name to use for the recorded bag file"/>
  <arg name="timestamp_mode" default=" " doc="A parameter that allows you to override the timestamp measurements;
    possible values: {
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="viz" default="true" doc="whether to run a rviz"/>
  <arg name="rviz_config" default="$(find ouster_ros)/config/viz.rviz" doc="optional rviz config file"/>

  <arg name="tf_prefix" default=" " doc="namespace for tf transforms"/>
  <arg name="sensor_frame" default="os_sensor"
    doc="sets name of choice for the sensor_frame tf frame, value can not be empty"/>
  <arg name="lidar_frame" default="os_lidar"
    doc="sets name of choice for the os_lidar tf frame, value can not be empty"/>
  <arg name="imu_frame" default="os_imu"
    doc="sets name of choice for the os_imu tf frame, value can not be empty"/>
  <arg name="point_cloud_frame" default=" "
    doc="which frame to be used when publishing PointCloud2 or LaserScan messages.
    Choose between the value of sensor_frame or lidar_frame, leaving this value empty
    would set lidar_frame to be the frame used when publishing these messages."/>

  <arg name="dynamic_transforms_broadcast" default="false"
    doc="static or dynamic transforms broadcast"/>
  <arg name="dynamic_transforms_broadcast_rate" default="1.0"
    doc="set the rate (Hz) of broadcast when using dynamic broadcast; minimum value is 1 Hz"/>

  <arg name="proc_mask" default="IMG|PCL|IMU|SCAN" doc="
    The IMG flag here is not supported and does not affect anything,
    to disable image topics you would need to omit the os_image node
    from the launch file"/>

  <arg name="scan_ring" default="0" doc="
    use this parameter in conjunction with the SCAN flag
    and choose a value the range [0, sensor_beams_count)"/>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_nodelet_mgr"
      output="screen" required="true"
      launch-prefix="bash -c 'sleep 2; $0 $@' "
      args="manager"/>
  </group>

  <arg name="_use_metadata_file" value="$(eval not (metadata == ''))"/>

  <group ns="$(arg ouster_ns)">
    <node if="$(arg _use_metadata_file)" pkg="nodelet" type="nodelet"
      name="os_node" output="screen" required="true"
      launch-prefix="bash -c 'sleep 3; $0 $@' "
      args="load ouster_ros/OusterReplay os_nodelet_mgr">
      <param name="~/metadata" value="$(arg metadata)"/>
    </node>
  </group>

  <include file="$(find ouster_ros)/launch/common.launch">
    <arg name="ouster_ns" value="$(arg ouster_ns)"/>
    <arg name="viz" value="$(arg viz)"/>
    <arg name="rviz_config" value="$(arg rviz_config)"/>
    <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    <arg name="sensor_frame" value="$(arg sensor_frame)"/>
    <arg name="lidar_frame" value="$(arg lidar_frame)"/>
    <arg name="imu_frame" value="$(arg imu_frame)"/>
    <arg name="point_cloud_frame" value="$(arg point_cloud_frame)"/>
    <arg name="timestamp_mode" value="$(arg timestamp_mode)"/>
    <arg name="dynamic_transforms_broadcast"
      value="$(arg dynamic_transforms_broadcast)"/>
    <arg name="dynamic_transforms_broadcast_rate"
      value="$(arg dynamic_transforms_broadcast_rate)"/>
    <arg name="proc_mask" value="$(arg proc_mask)"/>
    <arg name="scan_ring" value="$(arg scan_ring)"/>
  </include>

  <arg name="_use_bag_file_name" value="$(eval not (bag_file == ''))"/>

  <node if="$(arg _use_bag_file_name)" pkg="rosbag" type="play" name="rosbag_play_recording"
    launch-prefix="bash -c 'sleep 3; $0 $@' "
    output="screen" required="true"
    args="--clock $(arg bag_file)"/>

</launch>
